lib/libc/sys/varsym.2
.\" Copyright (c) 2008 The Aerospace Corporation
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the author nor the names of its contributors may
.\"    be used to endorse or promote products derived from this software
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD$
.\"
.Dd March 29 2008
.Dt VARSYM 2
.Os
.Sh NAME
.Nm varsym
.Nd manipulate variant symlink variables
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In varsym.h
.Ft int
.Fn varsym_set "int scope" "id_t which" "const char *name" "const char *data"
.Ft int
.Fn varsym_get "int scope" "id_t which" "const char *name" "const char *buf" "size_t *size"
.Ft int
.Fn varsym_list "int scope" "id_t which" "const char *buf" "size_t *size"
.Sh DESCRIPTION
The
.Fn varsym_set
function
sets the variable
.Fa name
in the object specifed by
.Fa scope
and
.Fa which
to the value in the null-terminated string pointed to by
.Fa data .
If
.Fa data
is
.Dv NULL
then the variable will be deleted.
If
.Fa name
is
.Dv NULL
then all variables will be deleted.
.Pp
The
.Fn varsym_get
function
writes the value of the variable
.Fa name
in the object specifed by
.Fa scope
and
.Fa which
to the buffer
.Fa buf
as a null-terminated string.
The number of bytes written to
.Fa which
are copied back to the variable pointed to by
.Fa size.
.Pp
The
.Fn varsym_list
function
writes the names and values of variables
in the object specifed by
.Fa scope
and
.Fa which
to the buffer
.Fa buf
with nulls between each entry and a final null indicating end of list.
The number of bytes written to
.Fa which
are copied back to the variable pointed to by
.Fa size.
.Pp
The
.Fa scope
argument may be one of:
.Bl -tag -width VARSYM_PROC
.It Dv VARSYM_SYS
Get or set variables in the global system scope.
.It Dv VARSYM_PROC
Get or set variables for the given process.
.El
.Pp
The following
.Xr sysctl 3
MIB variables control the operation of variant symlinks:
.Bl -tag -width vfs.varsym.enable
.It Va vfs.varsym.enable
Enables support for variant symlinks when set to a non-zero value.
Defaults to off.
.It Va vfs.varsym.allow_default
If set to a non-zero value, variant symlinks may be created with default
values to be used if no variable is set.
These values are specified by placing a colon after the variable name in the
link.
For example
.Va ${VAR:defaultvalue} .
.Pp
.Sy WARNING :
Default values should not be used in any way that will effect the execution
of setuid programs unless users can not control the values of variant
symlinks.
Administrators are advised to disable default value support when user
control of per-process values is allowed.
.It Va vfs.varsym.max_proc_setsize
Limits the number of per-processes variables that can be set.
The default limit is 32.
This limit is only enforced if unprivleged control is allowed by
.Va security.bsd.unprivileged_varsym_set_proc .
.It Va security.bsd.unprivileged_varsym_set_proc
If set to a non-zero value, non-root users may set per-process variant symlink
variables.
.El
.Sh RETURN VALUES
.Rv -std varsym_set varsym_get varsym_list
.Sh ERRORS
The
.Fn varsym_set ,
.Fn varsym_get ,
and
.Fn varsym_list
functions will fail and take no action if:
.Bl -tag -width Er
.It Bq Er EINVAL
An invalid value is passed for
.Fa scope
or
.Fa which .
.It Bq Er EPERM
The
.Fa which
variable is set to a value other than 0 or the current processes pid when
.Fa scope
is
.Dv VARSYM_PROC .
.It Bq Er EOVERFLOW
The buffer passed to
.Fn varsym_get
or
.Fn varsym_list
is too small for the data to be returned.
.It Bq Er E2BIG
For the
.Fn varsym_set
function, the maximum number of variant symlinks for this object has been
reached.
.It Bq Er ENOENT
The requested variant symlink variable does not exist.
.It Bq Er ENOSYS
The system does not support variant symlinks or the support has been
disabled by setting the
.Xr sysctl 3
MIB
.Va vfs.varsym.enable
to 0.
.It Bq Er EFAULT
One of buffers is outside the process's allocated address space.
.It Bq Er ENAMETOOLONG
The value of
.Fa name
is longer than
.Dv MAXVARSYM_NAME
or
the value of
.Fa data
is longer than
.Dv MAXVARSYM_DATA .
.El
.Sh SEE ALSO
.Xr varsym_clear 3
.Sh HISTORY
This version of variant symlinks is derived from variant symlink support that
first appeared in DragonFlyBSD 1.0.
.Sh BUGS
The
.Fa which
argument is effectivly unused at this time and must be 0 or the current
processes pid.


usr.bin/varsym/varsym.c
/*
 * Copyright (c) 2003,2004 The DragonFly Project.  All rights reserved.
 * Copyright (c) 2007-2009 The Aerospace Corporation.  All rights reserved.
 *
 * This code is derived from software contributed to The DragonFly Project
 * by Matthew Dillon <dillon@backplane.com>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 3. Neither the name of The DragonFly Project nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific, prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD$");

#include <sys/types.h>
#include <sys/varsym.h>
#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

static void listvars(int type);
static void printvar(int scope, const char* name);
static void usage(void);

int
main(int argc, char **argv)
{
	char *p;
	int ch;
	int scope;
	int printenv_mode;
	int want_clear, want_delete, want_proc, want_proc_priv, want_system;

	printenv_mode = 0;
	want_clear = want_delete = 0;
	want_proc = want_proc_priv = want_system = 0;
	while ((ch = getopt(argc, argv, "deiPps")) != -1)
		switch(ch) {
		case 'd':
			want_delete = 1;
			break;
		case 'e':
			printenv_mode = 1;
			break;
		case 'i':
			want_clear = 1;
			break;
		case 'P':
			want_proc_priv = 1;
			break;
		case 'p':
			want_proc = 1;
			break;
		case 's':
			want_system = 1;
			break;
		case '?':
		default:
			usage();
		}
	if (want_proc && want_proc_priv)
		errx(EXIT_FAILURE, "-p and -P are incompatable");
	if (want_proc && want_system)
		errx(EXIT_FAILURE, "-p and -s are incompatable");
	if (want_proc_priv && want_system)
		errx(EXIT_FAILURE, "-P and -s are incompatable");
	if (want_clear && want_delete)
		errx(EXIT_FAILURE, "-d and -i are incompatable");
	if (printenv_mode && want_delete)
		errx(EXIT_FAILURE, "-d and -P are incompatable");
	if (printenv_mode && want_clear)
		errx(EXIT_FAILURE, "-i and -P are incompatable");

	if (want_proc)
		scope = VARSYM_PROC;
	else if (want_proc_priv)
		scope = VARSYM_PROC_PRIV;
	else if (want_system)
		scope = VARSYM_SYS;
	else
		scope = VARSYM_ALL;

	argv += optind;
	if (*argv == NULL) {
		if (!want_clear) {
			listvars(scope);
			exit(0);
		} else if (scope != VARSYM_SYS) {
			errx(EXIT_FAILURE,
			    "-i requires a command or -s");
		}
		/*
		 * If we want to clear the system set, we fall through and
		 *  do that from here.
		 */
	}

	if (printenv_mode) {
		if (argv[1] != NULL)
			errx(EXIT_FAILURE, "Too many arguments.");
		printvar(scope, *argv);
	}

	/*
	 * If we're not listing variables, we actually want to default to
	 * process scope.
	 */
	if (scope == VARSYM_ALL)
		scope = VARSYM_PROC;

	if (want_delete) {
		for (; *argv != NULL; ++argv) {
			if (varsym_set(scope, 0, *argv, NULL) != 0 &&
			    errno != ENOENT)
				warn("failed to delete '%s'", *argv);
		}
		exit(0);
	}

	if (want_clear)
		if (varsym_set(scope, 0, NULL, NULL) != 0)
			err(EXIT_FAILURE, "varsym_clear()");

	for (; *argv && (p = strchr(*argv, '=')); ++argv) {
		*p = '\0';
		if (varsym_set(scope, 0, *argv, ++p) != 0)
			err(EXIT_FAILURE, "varsym_set(%s)", *argv);
		*p = '=';
	}

	if (*argv != NULL) {
		if (want_system)
			errx(EXIT_FAILURE, "too many arguments for -s");
		execvp(*argv, argv);
		err(errno == ENOENT ? 127 : 126, "%s", *argv);
	}

	exit(0);
}

static void
usage(void)
{
	(void)fprintf(stderr,
	    "usage: varsym [-ips] [name=value ...] [utility [argument ...]]\n"
	    "       varsym -d [-ps] [name ...]\n");
	exit(1);
}

static int
isinbuf(char *name, char *buf, int bytes)
{
	int b;
	int i;
	char *vname = NULL;

	for (b = i = 0; i < bytes; i++) {
		if (buf[i] == '\0') {
			if (vname == NULL) {
				vname = buf + b;
				if (strcmp(vname, name) == 0)
					return(1);
			} else {
				vname = NULL;
			}
			b = i + 1;
		}
	}
	return(0);
}

static void
dumpvars(char *buf, int bytes)
{
	int b;
	int i;
	char *vname = NULL;
	char *vdata = NULL;

	for (b = i = 0; i < bytes; i++) {
		if (buf[i] == 0) {
			if (vname == NULL) {
				vname = buf + b;
			} else {
				vdata = buf + b;
				if (!isinbuf(vname, buf, vname - buf))
					printf("%s=%s\n", vname, vdata);
				vname = vdata = NULL;
			}
			b = i + 1;
		}
	}
}

static void
listvars(int scope)
{
	char buf[3*128*1024]; /* XXX: MAXPHYS */
	size_t size, tmpsize;

	size = sizeof(buf);
	if (scope != VARSYM_ALL) {
		if (varsym_list(scope, 0, buf, &size) != 0)
			err(1, "varsym_list failed");
	} else {
		if (varsym_list(VARSYM_SYS, 0, buf, &size) != 0)
			err(1, "varsym_list failed");
		tmpsize = sizeof(buf) - size;
		if (varsym_list(VARSYM_PROC_PRIV, 0, buf+size, &tmpsize) != 0)
			err(1, "varsym_list failed");
		size += tmpsize;
		tmpsize = sizeof(buf) - size;
		if (varsym_list(VARSYM_PROC, 0, buf+size, &tmpsize) != 0)
			err(1, "varsym_list failed");
		size += tmpsize;
	}
	dumpvars(buf, size);
}

static void
printvar(int scope, const char* name)
{
	char buf[MAXVARSYM_DATA];
	size_t bufsize;

	bufsize = sizeof(buf);
	if (scope != VARSYM_ALL) {
		if (varsym_get(scope, 0, name, buf, &bufsize) != 0) {
			if (errno == ENOENT)
				exit(1);
			else
				err(EXIT_FAILURE, "varsym_get()");
		}
	} else {
		if (varsym_get(VARSYM_SYS, 0, name, buf, &bufsize) != 0) {
			if (errno != ENOENT)
				err(EXIT_FAILURE, "varsym_get()");
			if (varsym_get(VARSYM_PROC_PRIV, 0, name, buf,
			    &bufsize) != 0) {
				if (errno != ENOENT)
					err(EXIT_FAILURE, "varsym_get()");
				if (varsym_get(VARSYM_PROC, 0, name, buf,
				    &bufsize) != 0) {
					if (errno == ENOENT)
						exit(1);
					else
						err(EXIT_FAILURE,
						    "varsym_get()");
				}
			}
		}
	}
	printf("%s\n", buf);
	exit(0);
}
